src/app/clusters/general-diagnostics-server/general-diagnostics-server.cpp

```diff
@@ -202,126 +211,138 @@ CHIP_ERROR GeneralDiagosticsAttrAccess::Read(const ConcreteReadAttributePath & a
     }
+};
 
-    // Gets called when the device has been rebooted.
-    void OnDeviceRebooted(BootReasonType bootReason) override
-    {
-        ChipLogDetail(Zcl, "GeneralDiagnosticsDelegate: OnDeviceRebooted");
+GeneralDiagnosticsDelegate gDiagnosticDelegate;
 
-        ReportAttributeOnAllEndpoints(GeneralDiagnostics::Attributes::BootReasons::Id);
+} // anonymous namespace
 
-        // GeneralDiagnostics cluster should exist only for endpoint 0.
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace GeneralDiagnostics {
 
-        Events::BootReason::Type event{ bootReason };
-        EventNumber eventNumber;
+Server Server::instance;
 
-        CHIP_ERROR err = LogEvent(event, 0, eventNumber);
-        if (CHIP_NO_ERROR != err)
-        {
-            ChipLogError(Zcl, "GeneralDiagnosticsDelegate: Failed to record BootReason event: %" CHIP_ERROR_FORMAT, err.Format());
-        }
+/**********************************************************
+ * Server Implementation
+ *********************************************************/
+
+Server & Server::Instance()
+{
+    return instance;
+}
+
+// Gets called when the device has been rebooted.
+void Server::OnDeviceReboot(BootReasonType bootReason)
+{
+    ChipLogDetail(Zcl, "GeneralDiagnostics: OnDeviceReboot");
+
+    ReportAttributeOnAllEndpoints(GeneralDiagnostics::Attributes::BootReasons::Id);
+
+    // GeneralDiagnostics cluster should exist only for endpoint 0.
+
+    Events::BootReason::Type event{ bootReason };
+    EventNumber eventNumber;
+
+    CHIP_ERROR err = LogEvent(event, 0, eventNumber);
+    if (CHIP_NO_ERROR != err)
+    {
+        ChipLogError(Zcl, "GeneralDiagnostics: Failed to record BootReason event: %" CHIP_ERROR_FORMAT, err.Format());
     }
+}
+
+// Get called when the Node detects a hardware fault has been raised.
+void Server::OnHardwareFaultsDetect(GeneralFaults<kMaxHardwareFaults> & previous, GeneralFaults<kMaxHardwareFaults> & current)
+{
+    ChipLogDetail(Zcl, "GeneralDiagnostics: OnHardwareFaultsDetect");
 
-    // Get called when the Node detects a hardware fault has been raised.
-    void OnHardwareFaultsDetected(GeneralFaults<kMaxHardwareFaults> & previous,
-                                  GeneralFaults<kMaxHardwareFaults> & current) override
+    for (auto endpointId : EnabledEndpointsWithServerCluster(GeneralDiagnostics::Id))
     {
-        ChipLogDetail(Zcl, "GeneralDiagnosticsDelegate: OnHardwareFaultsDetected");
+        // If General Diagnostics cluster is implemented on this endpoint
+        MatterReportingAttributeChangeCallback(endpointId, GeneralDiagnostics::Id,
+                                               GeneralDiagnostics::Attributes::ActiveHardwareFaults::Id);
+
+        // Record HardwareFault event
+        EventNumber eventNumber;
+        DataModel::List<const HardwareFaultType> currentList =
+            DataModel::List<const HardwareFaultType>(reinterpret_cast<const HardwareFaultType *>(current.data()), current.size());
```

Comment:
```suggestion
        DataModel::List<const HardwareFaultType> currentList(reinterpret_cast<const HardwareFaultType *>(current.data()), current.size());
```
Or use `auto`.  Yes, I know this was pre-existing....